### PR TITLE
[test_mgmtvrf] Improve the robustness of mgmt VRF testing

### PR DESCRIPTION
**Pull request by wangxin**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [x] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### What is the motivation for this PR?
The test_mgmtvrf.py script has some issues need to be fixed.
1. In teardown, command `config save` is executed on DUT after mgmt VRF configuration is removed. However, if something went wrong with removing mgmt VRF configuration, the mgmt VRF configuration will be saved to the `/etc/sonic/config_db.json` file. This could leave the system in unhealthy status and affect rest of the tests.
2. Some test cases assume that the DUT switch has Internet access. This is not always the case in lab.
3. Test case test_wget and test_curl are covering the same scenario. It's kind of duplicated.

#### How did you do it?
Changes:
* Removed test_wget. The scenario is already covered by test_curl.
* Use temporary http server on PTF for test_curl. No need for DUT to have internet access.
* Backup and restore the original config_db in reboot related tests.
* Do not run 'config save' in teardown to avoid accidentially storing mgmt vrf in config_db and affecting other tests.
* Replaced root logger with module logger for logging.
* Moved the script to mvrf subfolder.

#### How did you verify/test it?
Test run the test_mgmtvrf.py script on latest 201911 image. All cases passed.

### CODE REVIEW

File: tests/mvrf/temp_http_server.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import sys
+
+from BaseHTTPServer import HTTPServer
+from BaseHTTPServer import BaseHTTPRequestHandler
+
+MAGIC_STRING = "MAGCI_STRING_FOR_TESTING"
```
Comment (daall) [inline tests/mvrf/temp_http_server.py:6]:
```suggestion
MAGIC_STRING = "MAGIC_STRING_FOR_TESTING"
```
typo?

File: tests/mvrf/test_mgmtvrf.py [hunk 2]
```diff
@@ -172,20 +174,28 @@ def basic_check_after_reboot(self, duthost, localhost, testbed):
...
 
-    def test_warmboot(self, duthost, localhost, testbed):
-        duthost.command("sudo config save -y")
+    @pytest.fixture(scope="class")
+    def backup_restore_config(self, duthost):
+        # Backup the original config_db without mgmt vrf config
+        duthost.shell("cp /etc/sonic/config_db.json /etc/sonic/config_db.json.bak")
+        yield
+        # Restore the original config_db to override the config_db with mgmt vrf config
+        duthost.shell("mv /etc/sonic/config_db.json.bak /etc/sonic/config_db.json")
```
Comment (daall) [inline tests/mvrf/test_mgmtvrf.py:183]: Do we need to do a `config reload` to restore the original config_db once the test is over?

File: tests/mvrf/temp_http_server.py (hunk 1 above)
Comment (wangxin) [reply tests/mvrf/temp_http_server.py]: Thanks! Fixed.

File: tests/mvrf/test_mgmtvrf.py (hunk 2 above)
Comment (wangxin) [reply tests/mvrf/test_mgmtvrf.py:190]: Sure. Pushed a new commit:
* Added config reload in teardown
* Added back the code for getting free_port, which was accidentally removed while creating PR.

Comment (yxieca) [reply tests/mvrf/test_mgmtvrf.py:190]: Would there be race condition between config restore and config reload?

Even there is no race, I think putting them together could prevent future human error.

Comment (wangxin) [reply tests/mvrf/test_mgmtvrf.py]: Updated. I also changed `pytest_assert` to `raise Exception` in the `verify_show_command` function. Then we can have a exception to catch if something is wrong. Otherwise, pytest_assert may just fail the test and there is no exception to catch. This would make it difficult to insert the logic of restoring config_db.
Now, if configuring mgmt vrf failed somehow, the exception can always be caught. The code for restoring config_db won't b missed. The pytest.fail after that will just fail the whole test and stop executing rest of the test cases and teardown part.

File: tests/mvrf/temp_http_server.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import sys
+
+from BaseHTTPServer import HTTPServer
+from BaseHTTPServer import BaseHTTPRequestHandler
```
Comment (daall) [inline tests/mvrf/temp_http_server.py:4]: Can we combine these two?
```suggestion
from BaseHTTPServer import HTTPServer, BaseHTTPRequestHandler
```

File: tests/mvrf/test_mgmtvrf.py [hunk 4]
```diff
@@ -14,60 +15,81 @@
...
-    SONIC_SSH_REGEX = "OpenSSH_[\\w\\.]+ Debian"
-
-    verify_show_command(duthost)
+    # Backup the original config_db without mgmt vrf config
+    duthost.shell("cp /etc/sonic/config_db.json /etc/sonic/config_db.json.bak")
+
+    try:
+        logger.info("Configure mgmt vrf")
+        global var
+        global mvrf
```
Comment (daall) [inline tests/mvrf/test_mgmtvrf.py:37]: I think this is leftover from before, so it probably makes sense to do this in a separate PR, but I think we should probably avoid using globals and instead `yield` these vars to the test cases that need them.

File: tests/mvrf/test_mgmtvrf.py [hunk 5]
```diff
@@ -14,60 +15,81 @@
...
+        var["dut_ip"] = duthost.setup()["ansible_facts"]["ansible_eth0"]["ipv4"]["address"]
+        var["ptf_ip"] = testbed["ptf_ip"]
+        var["filename"] = "README.md"
+        duthost.command("sudo config vrf add mgmt")
+        SONIC_SSH_REGEX = "OpenSSH_[\\w\\.]+ Debian"
+        verify_show_command(duthost)
+    except Exception as e:
+        logger.error("Exception raised in setup, exception: {}".format(repr(e)))
+        restore_config_db(duthost)
+        pytest.fail("Configure mgmt vrf failed, no test case will be executed. Teardown will not be executed either.")
```
Comment (daall) [inline tests/mvrf/test_mgmtvrf.py:49]: I don't know if this is accurate. We restored config DB, which is sort of teardown, right? :P

File: tests/mvrf/test_mgmtvrf.py (hunk 2 above)
Comment (daall) [reply tests/mvrf/test_mgmtvrf.py]: Yeah I think that looks good!

File: tests/mvrf/temp_http_server.py (hunk 3 above)
Comment (wangxin) [reply tests/mvrf/temp_http_server.py]: Updated.

File: tests/mvrf/test_mgmtvrf.py (hunk 4 above)
Comment (wangxin) [reply tests/mvrf/test_mgmtvrf.py:37]: Totally agree with you. We can address this in separate PR.

File: tests/mvrf/test_mgmtvrf.py (hunk 5 above)
Comment (wangxin) [reply tests/mvrf/test_mgmtvrf.py]: Right. Updated to "Code after 'yield' will not be executed either".

**Comment by irene-pan1202** [comment]
@wangxin 
After unconfiguring mgmt vrf, Why need to confirm that the SSH state has stopped? There is no such confirmation in the old version. Since it only took 1 second for some machines to disconnect, adding this confirmation seems to cause the test error. Thanks.
